src/AFQMC/Estimators/FullObsHandler.hpp

```diff
@@ -0,0 +1,241 @@
+  {
+    if(iav < 0 || iav >= nave)
+      APP_ABORT("Runtime Error: iav out of range in full1rdm::accumulate. \n\n\n");
+
+    int nw(wset.size());
+    int nrefs(Refs.size(1));
+    set_buffer( nw * (dm_size+3) );
+    sharedC4Tensor_ref G4D(Buff.origin(), {nw, nspins, std::get<0>(Gdims),std::get<1>(Gdims)});
+    sharedCMatrix_ref G2D(Buff.origin(), {nw, dm_size});
+    sharedCVector_ref DevOv(G4D.origin()+G4D.num_elements(), {nw});
+
+    stdCVector Xw(iextensions<1u>{nw});
+    stdCVector Ov(iextensions<1u>{nw});
+    stdCMatrix detR(DevdetR); 
+    std::vector<devCMatrix_ref> RefsA;
+    std::vector<devCMatrix_ref> RefsB;
+    RefsA.reserve(nw);
+    if(walker_type == COLLINEAR) RefsB.reserve(nw);
+
+    if(impsamp) 
+      denominator[iav] += std::accumulate(wgt.begin(),wgt.end(),ComplexType(0.0));
+    else {
+      APP_ABORT(" Finish implementation of free projection. \n\n\n");
+    }
+
+    for(int iref=0, is=0; iref<nrefs; iref++, is+=nspins) {
+
+      ComplexType CIcoeff(wfn0.getReferenceWeight(iref));
+
+      //1. Calculate Green functions
+      // Refs({wset.size(),nrefs,ref_size}  
+      RefsA.clear();  
+      RefsB.clear();  
+// NOTE: Incomplete, must copy to device memory.
+//       Use the Aux SM in the walker and then make array_refs from them. 
+      if(walker_type == CLOSED) {
+        for(int iw=0; iw<nw; iw++) 
+// push SMAux and them copy_n from Refs to there
+          RefsA.emplace_back(devCMatrix_ref(make_device_ptr(Refs[iw][iref].origin()), 
+                                            {NMO,NAEA}));
+      } else if(walker_type == COLLINEAR) {
+        for(int iw=0; iw<nw; iw++) { 
+          RefsA.emplace_back(devCMatrix_ref(make_device_ptr(Refs[iw][iref].origin()), 
+                                            {NMO,NAEA}));
+          RefsB.emplace_back(devCMatrix_ref(make_device_ptr(Refs[iw][iref].origin())+NAEA*NMO, 
+                                            {NMO,NAEB}));
+        }
+      } else if(walker_type == NONCOLLINEAR) {
+        for(int iw=0; iw<nw; iw++) 
+          RefsA.emplace_back(devCMatrix_ref(make_device_ptr(Refs[iw][iref].origin()), 
+                                            {2*NMO,NAEA+NAEB}));
+      }
+      wfn0.DensityMatrix(wset, RefsA, RefsB, G2D, DevOv, false, false, true);
+
+      //2. calculate and accumulate appropriate weights 
+      copy_n( make_device_ptr(DevOv.origin()), nw, Ov.origin());
+      if(walker_type == CLOSED) { 
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][iref] * detR[iw][iref]; 
+      } else if(walker_type == COLLINEAR) {
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][2*iref] * detR[iw][2*iref+1]; 
+      } else if(walker_type == NONCOLLINEAR) {
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][iref]; 
+      } 
+
+      //3. accumulate references 
```

Comment:
What is a reference in this context?

Reply:
A reference is a term in the (possibly) MSD expansion being back propagated.
Each determinant needs to be back propagated independently and the average
done at the end. References stands for "Reference Determinants" or 
"Determinants in the Trial Wavefunction"